File: android/versioned-abis/expoview-abi42_0_0/src/main/java/abi42_0_0/expo/modules/medialibrary/CreateAlbum.java
package abi42_0_0.expo.modules.medialibrary;

import android.content.Context;
import android.media.MediaScannerConnection;
import android.os.AsyncTask;
import android.provider.MediaStore;

import abi42_0_0.org.unimodules.core.Promise;

import java.io.File;
import java.io.IOException;
import java.util.List;

import static abi42_0_0.expo.modules.medialibrary.MediaLibraryConstants.ERROR_NO_ALBUM;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryConstants.ERROR_UNABLE_TO_LOAD;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryConstants.ERROR_UNABLE_TO_LOAD_PERMISSION;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryConstants.ERROR_UNABLE_TO_SAVE;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryUtils.FileStrategy;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryUtils.copyStrategy;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryUtils.getAssetsById;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryUtils.getEnvDirectoryForAssetType;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryUtils.moveStrategy;
import static abi42_0_0.expo.modules.medialibrary.MediaLibraryUtils.queryAlbum;

class CreateAlbum extends AsyncTask<Void, Void, Void> {
  private final Context mContext;
  private final Promise mPromise;
  private final String mAlbumName;
  private final String mAssetId;
  private final FileStrategy mStrategy;

  public CreateAlbum(Context context, String albumName, String assetId, boolean copyAsset, Promise promise) {
    mContext = context;
    mAlbumName = albumName;
    mAssetId = assetId;
    mPromise = promise;
    mStrategy = copyAsset ? copyStrategy : moveStrategy;
  }

  private File createAlbum(String mimeType) {
    File albumDir = getEnvDirectoryForAssetType(mimeType, false);
    if (albumDir == null) {
      mPromise.reject(ERROR_NO_ALBUM, "Could not guess asset type.");
      return null;
    }

    File album = new File(albumDir.getPath(), mAlbumName);
    if (!album.exists() && !album.mkdirs()) {
      mPromise.reject(ERROR_NO_ALBUM, "Could not create album directory.");
      return null;
    }
    return album;
  }

  @Override
  protected Void doInBackground(Void... params) {
    try {
      List<MediaLibraryUtils.AssetFile> files = getAssetsById(mContext, mPromise, mAssetId);
      if (files == null) {
        return null;
      }

      MediaLibraryUtils.AssetFile albumCreator = files.get(0);
      File album = createAlbum(albumCreator.getMimeType());
      if (album == null) {
        return null;
      }

      File newFile = mStrategy.apply(albumCreator, album, mContext);

      MediaScannerConnection.scanFile(
        mContext,
        new String[]{newFile.getPath()},
        null,
        (path, uri) -> {
          if (uri == null) {
            mPromise.reject(ERROR_UNABLE_TO_SAVE, "Could not add image to album.");
            return;
          }
          final String selection = MediaStore.Images.Media.DATA + "=?";
          final String[] args = {path};
          queryAlbum(mContext, selection, args, mPromise);
        });
    } catch (SecurityException e) {
      mPromise.reject(ERROR_UNABLE_TO_LOAD_PERMISSION,
        "Could not create album: need WRITE_EXTERNAL_STORAGE permission.", e);
    } catch (IOException e) {
      mPromise.reject(ERROR_UNABLE_TO_LOAD, "Could not read file or parse EXIF tags", e);
    }
    return null;
  }
}

File: packages/expo-media-library/android/src/main/java/expo/modules/medialibrary/albums/CreateAlbum.kt
package expo.modules.medialibrary.albums

import android.content.Context
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import expo.modules.medialibrary.AlbumException
import expo.modules.medialibrary.MediaLibraryUtils
import expo.modules.medialibrary.UnableToLoadException
import expo.modules.medialibrary.assets.CreateAssetWithAlbumFile
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import java.io.File
import java.io.IOException
import kotlin.coroutines.coroutineContext

suspend fun createAlbum(
  context: Context,
  albumName: String,
  assetId: String,
  copyAsset: Boolean
): Bundle? = withContext(Dispatchers.IO) {
  try {
    val mStrategy = if (copyAsset) AssetFileStrategy.copyStrategy else AssetFileStrategy.moveStrategy
    val files = MediaLibraryUtils.getAssetsById(context, assetId)
    val albumCreator = files[0]
    val album = createAlbumFile(albumCreator.mimeType, albumName)
    val newFile = mStrategy.apply(albumCreator, album, context)
    coroutineContext.ensureActive()

    val (path, uri) = MediaLibraryUtils.scanFile(context, arrayOf(newFile.path), null)
    coroutineContext.ensureActive()
    if (uri == null) {
      throw AlbumException("Could not add image to album.")
    }
    val selection = "${MediaStore.Images.Media.DATA}=?"
    val args = arrayOf(path)
    val bundle = queryAlbum(context, selection, args)
    return@withContext bundle
  } catch (e: SecurityException) {
    throw UnableToLoadException("Could not create album: need WRITE_EXTERNAL_STORAGE permission: ${e.message}", e)
  } catch (e: IOException) {
    throw UnableToLoadException("Could not read file or parse EXIF tags: ${e.message}", e)
  }
}

// Creates the album file and uses existing CreateAssetWithAlbumFile to create the asset inside the album.
suspend fun createAlbumWithInitialFileUri(context: Context, albumName: String, assetUri: Uri): Bundle? {
  val mimeType: String = MediaLibraryUtils.getMimeType(context.contentResolver, assetUri) ?: run {
    throw AlbumException("Failed to create album: could not determine MIME type of the asset with uri: `$assetUri`.")
  }
  val assetUriPath = assetUri.path ?: run {
    throw AlbumException("Failed to create album: could not determine path of the asset with uri: `$assetUri`.")
  }

  val albumFile: File = createAlbumFile(mimeType, albumName)
  val mediaFile: File = File(assetUriPath)

  if (!mediaFile.exists()) {
    throw AlbumException("Failed to create album: the local media file with uri: `$assetUri` does not exist.")
  }

  CreateAssetWithAlbumFile(context, assetUri.toString(), false, albumFile).execute()
  return getAlbum(context, albumName)
}
